pkg/minikube/bootstrapper/kubeadm/kubeadm.go

```diff
@@ -0,0 +1,354 @@
+	restoreTmpl := `
+	sudo kubeadm alpha phase certs all --config {{.KubeadmConfigFile}} &&
+	sudo /usr/bin/kubeadm alpha phase kubeconfig all --config {{.KubeadmConfigFile}} &&
+	sudo /usr/bin/kubeadm alpha phase controlplane all --config {{.KubeadmConfigFile}} &&
+	sudo /usr/bin/kubeadm alpha phase etcd local --config {{.KubeadmConfigFile}}
+	`
+	t := template.Must(template.New("restoreTmpl").Parse(restoreTmpl))
+
+	opts := struct {
+		KubeadmConfigFile string
+	}{
+		KubeadmConfigFile: constants.KubeadmConfigFile,
+	}
+
+	b := bytes.Buffer{}
+	if err := t.Execute(&b, opts); err != nil {
+		return err
+	}
+
+	if err := k.c.Run(b.String()); err != nil {
+		return errors.Wrapf(err, "running cmd: %s", b.String())
+	}
+
+	if err := restartKubeProxy(k8s); err != nil {
+		return errors.Wrap(err, "restarting kube-proxy")
+	}
+
+	return nil
+}
+
+func (k *KubeadmBootstrapper) SetupCerts(k8s bootstrapper.KubernetesConfig) error {
+	return bootstrapper.SetupCerts(k.c, k8s)
+}
+
+func (k *KubeadmBootstrapper) UpdateCluster(cfg bootstrapper.KubernetesConfig) error {
+	if cfg.ShouldLoadCachedImages {
+		// Make best effort to load any cached images
+		go machine.LoadImages(k.c, constants.GetKubeadmCachedImages(cfg.KubernetesVersion), constants.ImageCacheDir)
+	}
+	kubeadmCfg, err := k.generateConfig(cfg)
+	if err != nil {
+		return errors.Wrap(err, "generating kubeadm cfg")
+	}
+
+	files := []assets.CopyableFile{
+		assets.NewMemoryAssetTarget([]byte(kubeletService), constants.KubeletServiceFile, "0640"),
+		assets.NewMemoryAssetTarget([]byte(kubeletSystemdConf), constants.KubeletSystemdConfFile, "0640"),
+		assets.NewMemoryAssetTarget([]byte(kubeadmCfg), constants.KubeadmConfigFile, "0640"),
+	}
+
+	if err := addAddons(&files); err != nil {
+		return errors.Wrap(err, "adding addons to copyable files")
+	}
+
+	for _, f := range files {
+		if err := k.c.Copy(f); err != nil {
+			return errors.Wrapf(err, "transferring kubeadm file: %+v", f)
+		}
+	}
+	var g errgroup.Group
+	for _, bin := range []string{"kubelet", "kubeadm"} {
+		bin := bin
+		g.Go(func() error {
+			path, err := maybeDownloadAndCache(bin, cfg.KubernetesVersion)
+			if err != nil {
+				return errors.Wrapf(err, "downloading %s", bin)
+			}
+			f, err := assets.NewFileAsset(path, "/usr/bin", bin, "0641")
+			if err != nil {
+				return errors.Wrap(err, "making new file asset")
+			}
+			if err := k.c.Copy(f); err != nil {
+				return errors.Wrapf(err, "transferring kubeadm file: %+v", f)
+			}
+			return nil
+		})
+	}
+	if err := g.Wait(); err != nil {
+		return errors.Wrap(err, "downloading binaries")
+	}
+
+	err = k.c.Run(`
+sudo systemctl daemon-reload &&
+sudo systemctl enable kubelet &&
+sudo systemctl start kubelet
+`)
+	if err != nil {
+		return errors.Wrap(err, "starting kubelet")
+	}
+
+	return nil
+}
+
+func (k *KubeadmBootstrapper) generateConfig(k8s bootstrapper.KubernetesConfig) (string, error) {
+	t := template.Must(template.New("kubeadmConfigTmpl").Parse(kubeadmConfigTmpl))
```

Comment:
as a defensive measure, you could put this (and the other function-scoped `template.Must` calls above) at global scope, so that `minikube` panics on startup if any template becomes invalid

Reply:
That's a good idea. Will do